Allow in-place modification of coordinate BaseFrame and SkyCoord

taldcroft opened the pull request:
<!-- This comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our contributing guidelines,
https://github.com/astropy/astropy/blob/master/CONTRIBUTING.md .
Please be sure to check out our code of conduct,
https://github.com/astropy/astropy/blob/master/CODE_OF_CONDUCT.md . -->

<!-- If you are new or need to be re-acquainted with Astropy
contributing workflow, please see
http://docs.astropy.org/en/latest/development/workflow/development_workflow.html .
There is even a practical example at
https://docs.astropy.org/en/latest/development/workflow/git_edit_workflow_examples.html#astropy-fix-example . -->

<!-- Astropy coding style guidelines can be found here:
https://docs.astropy.org/en/latest/development/codeguide.html#coding-style-conventions
Our testing infrastructure enforces to follow a subset of the PEP8 to be
followed. You can check locally whether your changes have followed these by
running the following command:

flake8 astropy --count --select=E101,W191,W291,W292,W293,W391,E111,E112,E113,E30,E502,E722,E901,E902,E999,F822,F823

-->

<!-- Please just have a quick search on GitHub to see if a similar
pull request has already been posted.
We have old closed pull requests that might provide useful code or ideas
that directly tie in with your pull request. -->

<!-- We have several automatic features that run when a pull request is open.
They can appear daunting but do not worry because maintainers will help
you navigate them, if necessary. -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

This pull request is to address #6394, #5745 and #6348 by adding `__setitem__` to the `SkyCoord` and `BaseCoordinateFrame` classes to allow setting values in array-valued instances.

It also adds an `insert()` method and `SkyCoordInfo.new_like()` method to support table operations with coordinates (`vstack`, `dstack`, `insert`). With this update, `SkyCoord` is fully supported in `Table` for non-masked operations. With one small exception, the changes to the `table` package in this PR are only in testing. The table change was needed because `SkyCoord` does not allow reshaping by setting the `shape` attribute.

Key points:
- This is intentionally a limited implementation that requires the right hand side value to be entirely equivalent to the left side except for the actual representation data.  This includes strict class equality and all frame attributes being equivalent.
- This is the minimal requirement to support Table operations.
- Setting values in `Representation` objects is done by a private `_setitem` method so prevent updating representation data without clearing the parent frame cache.
- The key assumption here is that clearing the frame `cache` is always sufficient to prevent any inconsistencies after updating representation data. 

<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->

Closes #6394
Closes #5745
Closes #6348

taldcroft commented:
@adrn @eteq - still looking for some high-level feedback on this PR to allow a limited and safe version of setitem in coordinates.

taldcroft commented:
@adrn @eteq - still looking for some high-level feedback on this PR to allow a limited and safe version of setitem in coordinates.

mhvk commented on astropy/coordinates/sky_coordinate.py:
```diff
@@ -26,6 +27,63 @@
+    elif frame_attr_shape != frame.shape:
+        # How to broadcast a ShapedLikeNDArray subclass like Time. E.g.
+        # an obstime attribute with shape (3,) and the frame has shape (10, 3).
+        raise NotImplementedError('help me @mhvk!')
```
The attribute `__get__` should have ensured the attribute is either a scalar or has been broadcast to the frame shape: 
https://github.com/astropy/astropy/blob/60fb417789abd03269f4d3af412ee2004983e869/astropy/coordinates/attributes.py#L110-L114
But that of course may leave it unwriteable...  I guess the safest might be to do the same broadcasting (sadly, two different ways to do it until I get around to #8610) and then copy.

mhvk commented on astropy/coordinates/representation.py:
```diff
@@ -255,6 +255,19 @@ def _apply(self, method, *args, **kwargs):
                     apply_method(getattr(self, component)))
         return new
 
+    def _setitem(self, item, value):
```
My sense would be to move this method to `BaseFrame`, as I don't think `Representation` has to know about this.

Or implement `__setitem__` for representations and (somehow) take care that `coord.data` contains or returns a read-only version.

taldcroft replied:
This cannot easily be moved out to `BaseFrame` because of the differentials.  Notice that there are two versions of `_setitem` and the differentials one calls the `super()` version.  So these really are part of the `Representation` class structure (in that they deal with different behavior of subclasses).

On the other idea of making the representation `data` be read-only, that pretty much breaks everything, including the existing (as of 4.0) API for updating coordinates in place.

mhvk replied:
Hmm, fair point about the subclassing.

But I think I wasn't very clear about my alternative suggestion: it was to just define `__setitem__` with what you have here, since representations themselves have no reason to be immutable (they're just complicated quantities as far as I'm concerned, and your current `_setitem` ensures we do not set with a wrong representation class).

The read-only aspect I mentioned only because, as far as I can tell, the only reason representations are currently immutable is to ensure people do not mess up by setting data. But as you note, we in fact show work-arounds for that, so if we are making `SkyCoord` mutable anyway, perhaps we do not have to worry too much about `sky_coord.data` becoming mutable itself.

mhvk replied:
@eteq, @taldcroft - thinking more about this, I still think we should just give representations a `__setitem__` - they really have no reason to be immutable. On the caching question, I think it is important that we make sure our code works well, and that we have appropriate warnings to avoid user errors; I don't think we should make convoluted code just to ensure users cannot make mistakes. So, I would suggest to use `__setitem__` for the representations as well as for `SkyCoordinate` and `BaseFrame`, and be sure to add a warning to the docstring of `data`.

p.s. Just as an example of how hopeless it is to prevent users messing up if they go for attributes, consider a user who, quite logically in my opinion (arguably more logical at least than setting `data`), thinks that one can change coordinates from their named attributes:
```
sc = SkyCoord(np.arange(2.)*u.deg, np.arange(2.)*u.deg)                                        
sc                                                                                             
# <SkyCoord (ICRS): (ra, dec) in deg
#     [(0., 0.), (1., 1.)]>
sc.ra                                                                                          
# <Longitude [0., 1.] deg>
# Aha, I can just change RA, right?
sc.ra[1] = -1*u.deg                                                                            
# Let me just check to be sure
sc.ra                                                                                          
# <Longitude [  0., 359.] deg>
# Clearly, that worked.
# But...
sc                                                                                             
# <SkyCoord (ICRS): (ra, dec) in deg
#     [(0., 0.), (1., 1.)]>
```

taldcroft replied:
@mhvk - indeed the current infrastructure of frame / representation and the dependence on caching (for performance and user-experience reasons) leaves coordinates vulnerable to problems like your example.  That is an important separate discussion, but completely out of scope for this PR, which is just adding one safe way to set items.

What is the intended public-facing use-case for having `__setitem__` on representations?  Why do we need to expose that, and in the process expose an additional way of accidentally corrupting a coordinate object?

I don't think that calling a method `_setitem` vs. `__setitem__` changes the code complexity much.  I could move all of that code into `BaseFrame.__setitem__` but that would actually make the code more complex by making that method a lot bigger, with handling that is specific to the representation not the frame.

taldcroft replied:
I wonder if the conversation would be helped by describing the `.data` representation as "protected" instead of mutable or immutable.  The idea of "protected" is that users have to take unusual steps to update `.data` in place, in this case explicitly clearing a cache that is not directly part of the representation object itself.  The point of not providing `__setitem__` is to help the users to get to the documentation pages which describe the dragons.

mhvk replied:
I like "protected"!

My main reason for thinking representations should have `__setitem__` is that that is the standard way to set items in python, and if we allow setting in `SkyCoord` and the frames using `__setitem__` it just seems illogical not to do the same for representations. Why should the lower level be done in a different way? What makes them special?

I think this in particular since the representations are not only useful for coordinates, but also outside, as a general way to represent 3-d vectors (as we document; http://docs.astropy.org/en/latest/units/quantity.html#representing-vectors-with-units and http://docs.astropy.org/en/latest/coordinates/representations.html#vector-arithmetic). They much behave like quantities, so might as well be similarly settable.

mhvk replied:
p.s. Of course, none of this would matter if we didn't have to have caching... https://www.martinfowler.com/bliki/TwoHardThings.html

taldcroft replied:
>  Why should the lower level be done in a different way? What makes them special?

It is exactly the caching and the fact that `.data` does not have any reference to a parent frame, so it cannot invalidate the cache in `__setitem__`.  Caching is basically needed because of APE-5 and https://github.com/astropy/astropy-api/pull/12, e.g.
```
assert icrs.spherical.lat == 5*u.deg
assert icrs.cartesian.z.value > 0
icrs.representation_type = 'cartesian'  # It works
```
In other words the API allows seamlessly changing representations, so caching is needed to make that performant. I will accept some blame here, this seemed like a fine idea and the community accepted it.  If we had a time machine it would be worth trying to avoid anything in the API that requires caching.  But that ship has sailed.  I don't see a way out without make a new class or a serious API break.

Having said all this, can we open a new issue to "consider adding `__setitem__` to representations", and stipulate that this is out of scope for *this PR*, which is explicitly about making SkyCoord and Frame a little better and to support table operations for SkyCoord. In other words, perfect is the enemy of good.

mhvk replied:
Just to be clear, I am not claiming caching is a bad idea, just remarking that it is known in computer science that to do it right is very hard (Indeed, that has already been amply demonstrated for `Time`, where it similarly makes sense to cache, but where of course setting the equivalent of `.data`, `jd1[...]` and `jd2[...]` screws up the cache as well).

But before just giving in on making a separate issue (which just means I have to make the changes and we will need to have the whole discussion all over again): what is really wrong with just changing `_set_item` to `__setitem__`? Right now one can screw up nearly everything else in a `SkyCoord` by setting an item in most attributes (except `TimeAttribute`, which is made read-only; quantities, locations, and now coordinate origins are settable), so why make `.data` different. In fact, arguably doing so is confusing since not allowing setting it suggests it is safe to set all the other attributes.

And of course, by changing to `__setitem__`, table would work for representations as well! 

Note that there is *no* additional risk of previous code breaking; we are only talking about new users possibly messing up.

p.s. I'm quite willing to make a PR to yours that does it, adjusts the `.data` docstring, and adds some tests for representations.

taldcroft replied:
What is really the issue is that making the change is *controversial*, so I do not want my PR being blocked and delayed by a controversial discussion that has nothing to do with the goal of this PR.  That change is out of scope and so it really belongs in a separate PR.  We should keep each PR as small as possible, no?  

This one is done and ready to merge now (apart from one missing backtick and final review of doc updates from @eteq), and there is no reason it needs a new PR to make it complete or even better.  Once this is merged (which could happen as soon as @eteq looks over changes), then you can make a nice clean PR against master.

mhvk replied:
@taldcrof - OK, let's indeed go that route. I've gotten quite convinced this all is a good idea, and, as you say, we can do the next step in a separate PR.

mhvk commented:
The implementation here looks OK, but I guess the  general question is whether we are throwing away too much simplicity, and increasing the risk for inconsistency, by making `SkyCoord` mutable. I think if we stick to the current implementation where frames have to be identical, we are probably OK (it might have to be made slightly more flexible, like in the `Time.__setitem__` case, by explicitly checking that the item being replaced has exactly the same attributes, rather than the whole frame.).

But I do worry about the more general case, i.e., of setting an arbitrary `SkyCoord` with another arbitrary one (maybe including handling things like mixing proper motion with no proper motion, etc.). This will not be simple to get right. 

For instance, I do not think it is easy to keep the promise of in-place setting for the general case, in that while the coordinates themselves are set in place, in the background attributes may have to be copied (since they can be scalars but will need to have the coordinate's shape), which means things would not necessarily propagate back correctly. E.g., consider the following example
```
sc = SkyCoord([[1., 2.], [3., 4]]*u.hour, [[-10, -5], [5, 10]]*u.deg, obstime=Time('J2000'))
sc0 = sc[0]
sc0[1] = SkyCoord(5*u.hour, 20*u.deg, obstime=Time('J2010'))
```
And in any semi-trivial implementation the coordinates in `sc` will be changed, but `obstime` will not, because in `sc0` it will have to be copied (as it is a scalar) to allow setting an element.

It may be good to be clear about use cases. Presumably one of the case which this addresses is that in the documentation, where initializing a `SkyCoord` is too slow. But is the `is_equivalent_frame` check fast enough that this would actually help?

Another case is probably `Table`, in particular stacking and joining. For that case, would having a well-working stack/join option be good enough? I ask in particular since between clearing the cache and possibly updating attributes, one is effectively going to make quite a few copies anyway (just accessing `.ra` will make a copy in my above example cases, since those are in `hourangle`, and `.ra` is always in degrees), which suggests that perhaps ensuring that any change creates a new copy is the safest/easiest route. (I like a separate code path, e.g., implementing `np.stack` for coordinates, better since it isolates any problems...)

taldcroft commented:
>  I guess the general question is whether we are throwing away too much simplicity, and increasing the risk for inconsistency, by making SkyCoord mutable.

This PR does not make SkyCoord mutable, it has been mutable since at least astropy 2.0 (https://docs.astropy.org/en/stable/coordinates/inplace.html).  There are not any open issues (from what I could see) related to changing a SkyCoord in-place.  Even bing goes straight to the right documentation page for a search for "modify astropy skycoord in-place".  (Duckduckgo seemed lost in space however).

But certainly the goal here is to get the experts to try to conjure any scenario in which this PR (and the currently documented mutability) could result in a broken or inconsistent coordinate object.

My interest is largely related to making tables work across the board with astropy mixin columns, and to fix #6348 and #5745.  So I'm just thinking about supporting only a very limited case of precisely equivalent frames.  I'm not interested in the home-run of allowing arbitrary right-side coordinate attributes.  I agree that seems hard to impossible to do correctly all the time.

For high-performance one probably needs to use the back-door method that is currently documented.  But it is worth noting that doing a single in-place update with `__setitem__` here is about 3 times faster than making a new SkyCoord.  And we certainly could add a kwarg to `__setitem__` like `skip_checks` which allows skipping the checks if the caller is certain they are not needed.  For instance `sc.__setitem__(12, sc[13], skip_checks=True)`.

taldcroft commented:
A self-reminder to look back at #9536 which is related.

mhvk commented:
@taldcroft - I note that the current text at the link you provide states "Coordinates are generally considered to be immutable"...

But looking back at `Time`, where I was somewhat similarly worried, it has turned out that just providing the simple has-to-be-the-same location has been good enough. And similar strictness clearly is not very intrusive in terms of code. So, as long as we make clear in the documentation that we have no plans to generalize this, I'm fine with implementing your simple solution (i.e., without any change in attributes).

taldcroft commented:
> I note that the current text at the link you provide states "Coordinates are generally considered to be immutable"...

Good point, that needs to be changed!  :smirk_cat:

Are you also good with having a `validated` kwarg in `__setitem__` ala `TableColumns`?  This will help performance in table operations where we have just created the new SkyCoord with `info.new_like()` and already know that it is fine.  As usual, `new_like()` will take care that the inputs are equivalent.

https://github.com/astropy/astropy/blob/e85a7411ff124abab271f7f46e363a842fe18b7e/astropy/table/table.py#L252

mhvk commented:
Yes, :+1: to the additional keyword; easier than having to reimplement parts of `__setitem__` elsewhere.

taldcroft commented:
I started writing `new_like()` and realized that we may actually need `validated` (or something similar) instead of just wanting it.  Right now the `BaseCoordinateFrame.__setitem__` implementation ignores frame attributes entirely because it assumes they are all exactly the same. 

However, the idea of `new_like()` is that it makes a blank template object that gets filled in.  For a `SkyCoord` with an array attribute like `obstime` one needs to copy those array values in the same way as the primary data values, and that would happen in `__setitem__`.  Of course one could start with a more restricted limitation that does not allow table operations for a SkyCoord with array-valued frame attributes.

mhvk commented:
Yes, that makes sense.  It helps to be able to override things if one knows they are correct.

taldcroft commented:
@mhvk - this is ready for final review.  The docs build failure is just a whitespace error in the changelog.  I have that commit ready locally but I'll wait to push that pending your review.  The other two fails look unrelated.

taldcroft commented:
BTW I updated the top description as well.

taldcroft commented on astropy/coordinates/sky_coordinate.py:
```diff
@@ -83,6 +85,62 @@ def _represent_as_dict(self):
 
         return out
 
+    def new_like(self, cols, length, metadata_conflicts='warn', name=None):
```
Done.

taldcroft commented on astropy/coordinates/sky_coordinate.py:
```diff
@@ -356,6 +414,103 @@ def apply_method(value):
+    def __setitem__(self, item, value):
+        """Implement self[item] = value for SkyCoord
+
+        The right hand ``value`` must be strictly consistent with self:
```
Good point, done.

taldcroft commented on docs/coordinates/frames.rst:
```diff
@@ -252,6 +252,21 @@ set of coordinates, you will need to make sure that the shapes allow this::
    (as can be seen for ``obstime`` in the last line above).
 
+Coordinate values in a array-valued frame object can be modified in-place
+(added in astropy 4.1). This requires that the new values be set from an
```
Done

taldcroft commented on docs/coordinates/skycoord.rst:
```diff
@@ -357,6 +357,52 @@ use new views of the data, with the data copied only if that is impossible
+Modifying items in-place
+------------------------
+
+Coordinate values in a array-valued |SkyCoord| object can be modified in-place
```
I left this text here as the recommended no-dragons strategy and re-labeled that other document as "Fast In-place modification".  Along the way I put in a number of warnings callouts and generally updated the text to be more relevant and comprehensible (I hope!).

taldcroft commented on docs/coordinates/skycoord.rst:
```diff
@@ -357,6 +357,52 @@ use new views of the data, with the data copied only if that is impossible
   EXAMPLE END
 
+Modifying items in-place
+------------------------
```
Done.

taldcroft commented on astropy/coordinates/sky_coordinate.py:
```diff
@@ -83,6 +85,62 @@ def _represent_as_dict(self):
 
+    def new_like(self, cols, length, metadata_conflicts='warn', name=None):
+        """
+        Return a new SkyCoord instance which is consistent with the input
```
Done (clarified in docstring).

mhvk commented:
@taldcroft - was about to merge so I can base my `Representation` one on top, but there is a merge conflicts...  Could you rebase?

@eteq - last chance...

mhvk commented:
@eteq - p.s. Any comments on the documentation can obviously be addressed in the coming week; last chance is really about making coordinates mutable in the limited way proposed here (with `Representation`s coming as a follow-up by me)

mhvk commented:
@taldcroft - there is a problem: by default, since representations were meant to be immutable, input arrays are broadcast against each other, which means multiple items may be represented by the same memory. Hence, with the current PR:
```
In [7]: sc = SkyCoord(np.ones((3, 1)), np.ones((1, 2)), unit=u.deg)                                    

In [8]: sc                                                                                             
Out[8]: 
<SkyCoord (ICRS): (ra, dec) in deg
    [[(1., 1.), (1., 1.)],
     [(1., 1.), (1., 1.)],
     [(1., 1.), (1., 1.)]]>

In [9]: sc[1,1] = SkyCoord(0*u.deg, 0*u.deg)                                                           

In [10]: sc                                                                                            
Out[10]: 
<SkyCoord (ICRS): (ra, dec) in deg
    [[(1., 1.), (1., 0.)],
     [(0., 1.), (0., 0.)],
     [(1., 1.), (1., 0.)]]>
```

taldcroft commented:
Ugh changelog conflict. @bsipocz and @Cadair promised to make some wonderful system that would make this a thing of the past. :smile_cat:

taldcroft commented:
Rebased.  And of course git did the usual and screwed up the changelog in the rebase merge.

bsipocz commented:
> Ugh changelog conflict. @bsipocz and @Cadair promised to make some wonderful system that would make this a thing of the past. 

well, if we all on the team just did half that we planned to do 🤷

bsipocz commented:
@taldcroft - after rebase, please don't use the skip ci/travis etc tags. This still needs to run on travis before merging.

mhvk commented:
@taldcroft - the broadcasting is not entirely trivial to sort, as it is quite baked into the initializer: https://github.com/astropy/astropy/blob/8afea7252f73a916e889db90b2ac2f78c09c4a36/astropy/coordinates/representation.py#L149-L155

It may be OK to change the meaning of `copy` to include that a copy will be made if broadcasting is needed (for `copy=False`, people can be on their own...). My guess is that we'll need to revert the order: broadcast first, possibly copy, and then pass through the initializer with `copy=False`.

mhvk commented:
p.s. I thought I could just switch the broadcast and initializer calls, but it causes quite a number of errors. Some of those are specific tests that broadcasting works, so they don't really count, but others I do not quite understand.

mhvk commented:
OK, the problem is that by putting broadcasting first, one looses the ability to pass in lists of, e.g., longitude, since `np.array` does not concatenate these. Messy.

taldcroft commented:
> there is a problem: by default, since representations were meant to be immutable, input arrays are broadcast against each other

Argh. That seems like a bit of unnecessary optimization for how coordinates get used in real life (or maybe I'm missing some use case).  I really hope you can fix this, but maybe in the worst case find a way to raise an exception in setitem if the components are not the same shape?  Or just broadcast the components all to the same shape at that time (on-demand)?

mhvk commented:
It was simply the easiest way to ensure consistent shapes. I think now it should be possible to move the copy after the broadcast. I agree that it is not a particularly super-useful optimization since it gets lost the moment there is any transformation anyway.

mhvk commented:
@taldcroft - I added a commit which (I think) fixes things. Please have a look...

mhvk commented:
Oh, darn, I forgot to adjust the representation tests...

mhvk commented:
Should be OK now.

mhvk commented:
Darn, I think I have somehow removed one of the docs-fixing commits. Will try to fix tomorrow.

taldcroft commented:
In theory I have fixed it.  This is the result of git automerge being broken and actually corrupting the changelog, and it is now the second time I fixed that problem in this repo.

mhvk commented:
OK, thanks. Let's get this in!

taldcroft commented:
@mhvk - thanks for the review and good catch + fix on the broadcasting issue.

bsipocz commented:
This may deserve a what's new entry, as the skycoord mixin support for table operations is a great improvement
